pallets/runtime/src/bridge.rs

```diff
@@ -0,0 +1,303 @@
+//! Bridge from Ethereum to Polymesh
+//!
+//! This module implements a one-way bridge between Polymath Classic on the Ethereum side, and
+//! Polymesh native. It mints POLY on Polymesh in return for permanently locked ERC20 POLY tokens.
+
+use crate::multisig;
+use codec::{Decode, Encode};
+use frame_support::dispatch::{DispatchError, DispatchResult};
+use frame_support::traits::Currency;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, ensure, weights::GetDispatchInfo, Parameter,
+};
+use frame_system::{self as system, ensure_signed};
+use polymesh_primitives::{traits::IdentityCurrency, AccountKey, IdentityId, Signatory};
+use polymesh_runtime_balances as balances;
+use polymesh_runtime_common::{balances::Trait as BalancesTrait, CommonTrait};
+use polymesh_runtime_identity as identity;
+use sp_core::H256;
+use sp_runtime::traits::Dispatchable;
+use sp_std::collections::btree_map::BTreeMap;
+use sp_std::{convert::TryFrom, prelude::*};
+
+pub trait Trait: BalancesTrait + multisig::Trait {
+    type Balance: From<u128> + Into<<Self as CommonTrait>::Balance>;
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+    type Proposal: From<Call<Self>>
+        + Into<<Self as identity::Trait>::Proposal>
+        + Parameter
+        + Dispatchable<Origin = Self::Origin>
+        + GetDispatchInfo;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// The validator does not have an identity.
+        IdentityMissing,
+        /// Failure to credit the recipient account.
+        CannotCreditAccount,
+        /// Failure to credit the recipient identity.
+        CannotCreditIdentity,
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Bridge {
+        /// The multisig account of the bridge validator set.
+        Validators get(validators) config(): T::AccountId;
+        /// Correspondence between validator set change proposals and multisig proposal IDs.
+        ChangeValidatorsProposals get(change_validators_proposals): map T::AccountId => Option<u64>;
+        /// Correspondence between bridge transaction proposals and multisig proposal IDs.
+        BridgeTxProposals get(bridge_tx_proposals): map hasher(blake2_256) BridgeTx<T::AccountId> => Option<u64>;
```

Comment:
Rather than basing this on the whole `BridgeTx<T::AccountId>`, we can base this mapping on just the nonce inside the tx.